installing/installing_alibaba/installing-alibaba-customizations.adoc

```diff
@@ -0,0 +1,60 @@
+[id="installing-alibaba-customizations"]
+= Installing a cluster on Alibaba with customizations
+include::modules/common-attributes.adoc[]
+:context: installing-alibaba-customizations
+
+toc::[]
+
+In {product-title} version {product-version}, you can install a customized cluster on infrastructure that the installation program provisions on Alibaba Cloud. To customize the installation, you modify
+parameters in the `install-config.yaml` file before you install the cluster.
+
+[NOTE]
+====
+The scope of the {product-title} installation configurations is intentionally narrow. It is designed for simplicity and ensured success. You can complete many more {product-title} configuration tasks after an installation completes.
+====
+
+== Prerequisites
+
+* You reviewed details about the xref:../../architecture/architecture-installation.adoc#architecture-installation[{product-title} installation and update] processes.
+* You read the documentation on xref:../../installing/installing-preparing.adoc#installing-preparing[selecting a cluster installation method and preparing it for users].
+* You xref:../../installing/installing_alibaba/installing-alibaba-account.adoc#installing-alibaba-account[configured an Alibaba Cloud account] to host the cluster.
+* If you use a firewall, you xref:../../installing/install_config/configuring-firewall.adoc#configuring-firewall[configured it to allow the sites] that your cluster requires access to.
+* If the cloud identity and access management (IAM) APIs are not accessible in your environment, or if you do not want to store an administrator-level credential secret in the `kube-system` namespace, you can xref:../../installing/installing_alibaba/manually-creating-iam-alibaba.adoc#manually-creating-iam-alibaba[manually create and maintain IAM credentials].
```

Comment:
I am also not sure if we can install the cluster without CCO, granting one permission to all resources required to access the Cloud services.
@kwoodson @joelddiaz  could you please help with that topic?

Reply:
FYI CCO in manual mode does be needed, per the QE testing. And we were using the command line tool "ccoctl" to extract credential requests and create RAM users (Please refer to https://docs.openshift.com/container-platform/4.9/authentication/managing_cloud_provider_credentials/cco-mode-sts.html#cco-ccoctl-configuring_cco-mode-sts). Below are the 3 steps to deploy a cluster with CCO in manual mode. 

Step 1/3: create install-config and manifests

$ openshift-install create install-config --dir installation-dir
? SSH Public Key /home/fedora/.ssh/openshift-qe.pub
? Platform alibabacloud
? Region us-east-1
? Base Domain alicloud-qe.devcluster.openshift.com
? Cluster Name qe-cluster0126
? Pull Secret [? for help] ******
INFO Install-Config created in: installation-dir
$ echo 'credentialsMode: Manual' >> installation-dir/install-config.yaml
$ openshift-install create manifests --dir installation-dir
INFO Consuming Install Config from target directory 
INFO Manifests created in: installation-dir/manifests and installation-dir/openshift 
$ 

Step 2/3: extract 'ccoctl' and credentials-requests, then create-ram-users

$ export CCO_IMAGE=$(oc adm -a pull_secret.json release info --image-for='cloud-credential-operator' registry.ci.openshift.org/ocp/release:4.10.0-0.nightly-2022-01-25-023600)
$ echo ${CCO_IMAGE}
quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:f95da0a4f12e615150a74d967e1d0a80818880d448f15576a3814abfab5ef616
$ oc image extract ${CCO_IMAGE} --file="/usr/bin/ccoctl" -a pull_secret.json
$ chmod u+x ccoctl
$ ./ccoctl alibabacloud -h
Creating/deleting cloud credentials objects for alibaba cloud

Usage:
  ccoctl alibabacloud [command]

Available Commands:
  create-ram-users Create RAM Users and policies
  delete-ram-users Detach RAM Policy from existing user

Flags:
  -h, --help   help for alibabacloud

Use "ccoctl alibabacloud [command] --help" for more information about a command.
$ mkdir {cco-credrequests,cco-manifests}
$ oc adm -a pull_secret.json release extract --credentials-requests --cloud=alibabacloud --to="cco-credrequests" registry.ci.openshift.org/ocp/release:4.10.0-0.nightly-2022-01-25-023600
$ ls -l cco-credrequests
total 16
-rw-rw-r--. 1 fedora fedora  767 Jan 26 07:29 0000_30_machine-api-operator_00_credentials-request.yaml
-rw-rw-r--. 1 fedora fedora 2113 Jan 26 07:29 0000_50_cluster-image-registry-operator_01-registry-credentials-request-alibaba.yaml
-rw-rw-r--. 1 fedora fedora  763 Jan 26 07:29 0000_50_cluster-ingress-operator_00-ingress-credentials-request.yaml
-rw-rw-r--. 1 fedora fedora 1517 Jan 26 07:29 0000_50_cluster-storage-operator_03_credentials_request_alibaba.yaml
$ cat installation-dir/manifests/cluster-infrastructure-02-config.yml
apiVersion: config.openshift.io/v1
kind: Infrastructure
metadata:
  creationTimestamp: null
  name: cluster
spec:
  cloudConfig:
    key: config
    name: cloud-provider-config
  platformSpec:
    type: AlibabaCloud
status:
  apiServerInternalURI: https://api-int.qe-cluster0126.alicloud-qe.devcluster.openshift.com:6443
  apiServerURL: https://api.qe-cluster0126.alicloud-qe.devcluster.openshift.com:6443
  controlPlaneTopology: HighlyAvailable
  etcdDiscoveryDomain: ""
  infrastructureName: qe-cluster0126-kpbbn
  infrastructureTopology: HighlyAvailable
  platform: AlibabaCloud
  platformStatus:
    alibabaCloud:
      region: us-east-1
    type: AlibabaCloud
$ yq e '.status.infrastructureName' installation-dir/manifests/cluster-infrastructure-02-config.yml
qe-cluster0126-kpbbn
$ ./ccoctl alibabacloud create-ram-users --region us-east-1 --name qe-cluster0126-kpbbn --credentials-requests-
dir cco-credrequests --output-dir cco-manifests
2022/01/26 07:33:16 Created RAM User: qe-cluster0126-kpbbn-openshift-machine-api-alibabacloud-credenti
2022/01/26 07:33:16 Ready for creating new ram policy qe-cluster0126-kpbbn-openshift-machine-api-alibabacloud-credentials-policy-policy
2022/01/26 07:33:17 RAM policy qe-cluster0126-kpbbn-openshift-machine-api-alibabacloud-credentials-policy-policy has created
2022/01/26 07:33:17 Policy qe-cluster0126-kpbbn-openshift-machine-api-alibabacloud-credentials-policy-policy has attached on user qe-clu
ster0126-kpbbn-openshift-machine-api-alibabacloud-credenti
2022/01/26 07:33:18 Created access keys for RAM User: qe-cluster0126-kpbbn-openshift-machine-api-alibabacloud-credenti
2022/01/26 07:33:18 Saved credentials configuration to: cco-manifests/manifests/openshift-machine-api-alibabacloud-credentials-credentials.yaml
2022/01/26 07:33:18 Created RAM User: qe-cluster0126-kpbbn-openshift-image-registry-installer-cloud-cr
2022/01/26 07:33:18 Ready for creating new ram policy qe-cluster0126-kpbbn-openshift-image-registry-installer-cloud-credentials-policy-policy
2022/01/26 07:33:19 RAM policy qe-cluster0126-kpbbn-openshift-image-registry-installer-cloud-credentials-policy-policy has created
2022/01/26 07:33:19 Policy qe-cluster0126-kpbbn-openshift-image-registry-installer-cloud-credentials-policy-policy has attached on user qe-cluster0126-kpbbn-openshift-image-registry-installer-cloud-cr
2022/01/26 07:33:20 Created access keys for RAM User: qe-cluster0126-kpbbn-openshift-image-registry-installer-cloud-cr
2022/01/26 07:33:20 Saved credentials configuration to: cco-manifests/manifests/openshift-image-registry-installer-cloud-credentials-credentials.yaml
2022/01/26 07:33:20 Created RAM User: qe-cluster0126-kpbbn-openshift-ingress-operator-cloud-credential
2022/01/26 07:33:21 Ready for creating new ram policy qe-cluster0126-kpbbn-openshift-ingress-operator-cloud-credentials-policy-policy
2022/01/26 07:33:21 RAM policy qe-cluster0126-kpbbn-openshift-ingress-operator-cloud-credentials-policy-policy has created
2022/01/26 07:33:21 Policy qe-cluster0126-kpbbn-openshift-ingress-operator-cloud-credentials-policy-policy has attached on user qe-cluster0126-kpbbn-openshift-ingress-operator-cloud-credential
2022/01/26 07:33:22 Created access keys for RAM User: qe-cluster0126-kpbbn-openshift-ingress-operator-cloud-credential
2022/01/26 07:33:22 Saved credentials configuration to: cco-manifests/manifests/openshift-ingress-operator-cloud-credentials-credentials.yaml
2022/01/26 07:33:23 Created RAM User: qe-cluster0126-kpbbn-openshift-cluster-csi-drivers-alibaba-disk-
2022/01/26 07:33:23 Ready for creating new ram policy qe-cluster0126-kpbbn-openshift-cluster-csi-drivers-alibaba-disk-credentials-policy-policy
2022/01/26 07:33:23 RAM policy qe-cluster0126-kpbbn-openshift-cluster-csi-drivers-alibaba-disk-credentials-policy-policy has created
2022/01/26 07:33:24 Policy qe-cluster0126-kpbbn-openshift-cluster-csi-drivers-alibaba-disk-credentials-policy-policy has attached on user qe-cluster0126-kpbbn-openshift-cluster-csi-drivers-alibaba-disk-
2022/01/26 07:33:24 Created access keys for RAM User: qe-cluster0126-kpbbn-openshift-cluster-csi-drivers-alibaba-disk-
2022/01/26 07:33:24 Saved credentials configuration to: cco-manifests/manifests/openshift-cluster-csi-drivers-alibaba-disk-credentials-credentials.yaml
$ tree cco-manifests/
cco-manifests/
└── manifests
    ├── openshift-cluster-csi-drivers-alibaba-disk-credentials-credentials.yaml
    ├── openshift-image-registry-installer-cloud-credentials-credentials.yaml
    ├── openshift-ingress-operator-cloud-credentials-credentials.yaml
    └── openshift-machine-api-alibabacloud-credentials-credentials.yaml

1 directory, 4 files
$ cp cco-manifests/manifests/*.yaml installation-dir/manifests/
$ ls -l installation-dir/manifests/
total 88
-rw-r-----. 1 fedora fedora  623 Jan 26 07:27 cloud-provider-config.yaml
-rw-r-----. 1 fedora fedora 1285 Jan 26 07:27 cluster-config.yaml
-rw-r-----. 1 fedora fedora  373 Jan 26 07:27 cluster-dns-02-config.yml
-rw-r-----. 1 fedora fedora  667 Jan 26 07:27 cluster-infrastructure-02-config.yml
-rw-r-----. 1 fedora fedora  185 Jan 26 07:27 cluster-ingress-02-config.yml
-rw-r-----. 1 fedora fedora 9607 Jan 26 07:27 cluster-network-01-crd.yml
-rw-r-----. 1 fedora fedora  272 Jan 26 07:27 cluster-network-02-config.yml
-rw-r-----. 1 fedora fedora  142 Jan 26 07:27 cluster-proxy-01-config.yaml
-rw-r-----. 1 fedora fedora  171 Jan 26 07:27 cluster-scheduler-02-config.yml
-rw-r-----. 1 fedora fedora  200 Jan 26 07:27 cvo-overrides.yaml
-rw-r-----. 1 fedora fedora  118 Jan 26 07:27 kube-cloud-config.yaml
-rw-r-----. 1 fedora fedora 1304 Jan 26 07:27 kube-system-configmap-root-ca.yaml
-rw-r-----. 1 fedora fedora 4102 Jan 26 07:27 machine-config-server-tls-secret.yaml
-rw-------. 1 fedora fedora  292 Jan 26 07:34 openshift-cluster-csi-drivers-alibaba-disk-credentials-credentials.yaml
-rw-r-----. 1 fedora fedora 4181 Jan 26 07:27 openshift-config-secret-pull-secret.yaml
-rw-------. 1 fedora fedora  290 Jan 26 07:34 openshift-image-registry-installer-cloud-credentials-credentials.yaml
-rw-------. 1 fedora fedora  282 Jan 26 07:34 openshift-ingress-operator-cloud-credentials-credentials.yaml
-rw-------. 1 fedora fedora  284 Jan 26 07:34 openshift-machine-api-alibabacloud-credentials-credentials.yaml
$ 

Step 3/3: create cluster
$ openshift-install create cluster --dir installation-dir --log-level info

Reply:
> I am also not sure if we can install the cluster without CCO, granting one permission to all resources required to access the Cloud services.
> @kwoodson @joelddiaz could you please help with that topic?
You can have one shared set of credentials if that's what you want. In fact, that was the original implementation of support for AliCloud. But the `ccoctl` tool will not help you accomplish that. You would have to create all the Secret yaml files (for the installer) so that they all share the same access_key/secret_key data.

Reply:
> > I am also not sure if we can install the cluster without CCO, granting one permission to all resources required to access the Cloud services.
> > @kwoodson @joelddiaz could you please help with that topic?
> > You can have one shared set of credentials if that's what you want. In fact, that was the original implementation of support for AliCloud. But the `ccoctl` tool will not help you accomplish that. You would have to create all the Secret yaml files (for the installer) so that they all share the same access_key/secret_key data.

@kwoodson @joelddiaz @mburke5678 I am currently working on Azure Stack Hub IPI doc and it sounds like Alibaba Cloud shares the same requirement that the CCO be put into manual mode before creating the cluster, which includes creating the yaml files for the required secrets  [1]. I am wondering if the same content could be shared in the Alibaba doc. It is worth noting that the steps I mention do not instruct the user to create the install-config before creating the manifests. This is intentional in the case of ASH, as users must complete this step earlier in the process.

[1] [Manual manage cloud credentials](https://deploy-preview-39366--osdocs.netlify.app/openshift-enterprise/latest/installing/installing_azure_stack_hub/installing-azure-stack-hub-default.html#manually-create-iam_installing-azure-stack-hub-default)

Reply:
Yes, there is no need to create the install-config before the Secret yaml files are generated. You just need to create the Secret yaml files, and stop the installer at the right place to provide those Secret files before proceeding (along with setting the `credentialsMode: Manual` step).

Reply:
> Yes, there is no need to create the install-config before the Secret yaml files are generated. You just need to create the Secret yaml files, and stop the installer at the right place to provide those Secret files before proceeding (along with setting the `credentialsMode: Manual` step).

@joelddiaz What do you mean by stop the installer at the right place? For clarity - I am interpreting your statement as (1) create the manifests and update the yaml file for each `CredentialRequest` object; and (2) kick off the install. At what point is the user providing the updated yaml files that include the secrets?

Reply:
I mean: rather than run the full `openshift-install create cluster` you do it step-by-step by running `openshift-install create install-config` ; make some edits to the install-config ; `openshift-install create manifests` ; insert the extra manifests you need ; and finally `openshift-install create cluster`

Reply:
@joelddiaz Thank you. That makes sense now. One follow on to that. As I understand it, users can use the interactive prompts that the installer displays to select Alibaba and provide basic information like base domain, pull secret, etc. I am new to the install process, so please excuse the question if it is rudimentary. Presuming that the only value that is set in the install-config is `credentialsMode: Manual`, when the installer consumes the file, will the interactive prompts still be displayed?

Reply:
I've never walked through an AliCloud cluster installation end-to-end. So I'm not in the best position to answer this, but based on what I see, there are still several questions that you need to answer during the install interview to get an install-config that can then be edited:
```
[jdiaz@minigoomba os-install-4.10-fc1]$ ./openshift-install create install-config
? SSH Public Key /home/jdiaz/.ssh/id_rsa.pub
? Platform alibabacloud
? Region cn-hangzhou
? Base Domain jdiaz.ali.hive.openshift.com
? Cluster Name jdiaz-ali
? Pull Secret [? for help] ********************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************INFO Install-Config created in: .                 *********************************
[jdiaz@minigoomba os-install-4.10-fc1]$
```

Reply:
@joelddiaz Kenny [added the same to my PR](https://github.com/openshift/openshift-docs/pull/41083#discussion_r795764973). I [captured it here](https://github.com/openshift/openshift-docs/pull/41083/files#diff-aa2bf98d200d7bbc6fbc805b89a39df30e54112c533f035e025e1160cfeeecd9R211-R216).

Thanks!!